contracts/strategy/yearn/YearnStrategy.sol

```diff
@@ -0,0 +1,193 @@
+
+/**
+ * YearnStrategy generates yield by investing into a Yearn vault.
+ *
+ * @notice This strategy is syncrhonous (supports immediate withdrawals).
+ */
+contract YearnStrategy is IStrategy, AccessControl, CustomErrors {
+    using SafeERC20 for IERC20;
+    using PercentMath for uint256;
+    using ERC165Query for address;
+
+    // yearn vault is 0x
+    error StrategyYearnVaultCannotBe0Address();
+
+    /// role allowed to invest/withdraw from yearn vault
+    bytes32 public constant MANAGER_ROLE = keccak256("MANAGER_ROLE");
+    // underlying ERC20 token
+    IERC20 public immutable underlying;
+    /// @inheritdoc IStrategy
+    address public immutable override(IStrategy) vault;
+    // yearn vault that this strategy is interacting with
+    IYearnVault public immutable yVault;
+
+    /**
+     * @param _vault address of the vault that will use this strategy
+     * @param _owner address of the owner of this strategy
+     * @param _yVault address of the yearn vault that this strategy is using
+     * @param _underlying address of the underlying token
+     */
+    constructor(
+        address _vault,
+        address _owner,
+        address _yVault,
+        address _underlying
+    ) {
+        if (_owner == address(0)) revert StrategyOwnerCannotBe0Address();
+        if (_yVault == address(0)) revert StrategyYearnVaultCannotBe0Address();
+        if (_underlying == address(0))
+            revert StrategyUnderlyingCannotBe0Address();
+
+        if (!_vault.doesContractImplementInterface(type(IVault).interfaceId))
+            revert StrategyNotIVault();
+
+        _setupRole(DEFAULT_ADMIN_ROLE, _owner);
+        _setupRole(MANAGER_ROLE, _vault);
+
+        vault = _vault;
+        yVault = IYearnVault(_yVault);
+        underlying = IERC20(_underlying);
+    }
+
+    //
+    // Modifiers
+    //
+
+    modifier onlyManager() {
+        if (!hasRole(MANAGER_ROLE, msg.sender))
+            revert StrategyCallerNotManager();
+        _;
+    }
+
+    /**
+     * Yearn strategy is synchronous meaning it supports immediate withdrawals.
+     *
+     * @return true always
+     */
+    function isSync() external pure override(IStrategy) returns (bool) {
+        return true;
+    }
+
+    /// @inheritdoc IStrategy
+    function hasAssets()
+        external
+        view
+        virtual
+        override(IStrategy)
+        returns (bool)
+    {
+        return _getShares() > 0;
+    }
+
+    /// @inheritdoc IStrategy
+    function investedAssets()
+        external
+        view
+        virtual
+        override(IStrategy)
+        returns (uint256)
+    {
+        return _sharesToUnderlying(_getShares());
+    }
+
+    /// @inheritdoc IStrategy
+    function invest() external virtual override(IStrategy) onlyManager {
+        uint256 balance = _getUnderlyingBalance();
+        if (balance == 0) revert StrategyNoUnderlying();
+
+        underlying.safeIncreaseAllowance(address(yVault), balance);
```

Comment:
can't we just do a underlying.safeIncreaseAllowance MAX in the constructor and go with it? this is just transferring money between contracts cc @naps62

Reply:
yeah unless the yVault contract has an owner controlled sweep method or anything, we can do it

Reply:
👍